src/builtin_read.cpp

```diff
@@ -446,23 +462,42 @@ int builtin_read(parser_t &parser, io_streams_t &streams, wchar_t **argv) {
             if (i < argc) env_set(argv[i], &buff[j], opts.place);
         }
     } else if (opts.array) {
-        wcstring tokens;
-        tokens.reserve(buff.size());
-        bool empty = true;
-
-        for (wcstring_range loc = wcstring_tok(buff, ifs); loc.first != wcstring::npos;
-             loc = wcstring_tok(buff, ifs, loc)) {
-            if (!empty) tokens.push_back(ARRAY_SEP);
-            tokens.append(buff, loc.first, loc.second);
-            empty = false;
+        if (!opts.have_delimiter) {
+            // We're using IFS, so well split on every character in that,
+            // not on the entire thing at once.
+            wcstring tokens;
+            tokens.reserve(buff.size());
+            bool empty = true;
```

Comment:
I know you just moved this code but it seems like we can eliminate the `empty` var and just test `tokens.empty()`.